src/main/java/com/conveyal/analysis/datasource/derivation/AggregationAreaDerivation.java

```diff
@@ -0,0 +1,225 @@
+import spark.Request;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.zip.GZIPOutputStream;
+
+import static com.conveyal.file.FileStorageFormat.GEOJSON;
+import static com.conveyal.file.FileStorageFormat.SHP;
+import static com.conveyal.r5.analyst.WebMercatorGridPointSet.parseZoom;
+import static com.conveyal.r5.analyst.progress.WorkProductType.AGGREGATION_AREA;
+import static com.conveyal.r5.util.ShapefileReader.GeometryType.POLYGON;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Created by abyrd on 2021-09-03
+ */
+public class AggregationAreaDerivation implements DataDerivation<SpatialDataSource, AggregationArea> {
+
+    /**
+     * Arbitrary limit to prevent UI clutter from many aggregation areas (e.g. if someone uploads thousands of blocks).
+     * Someone might reasonably request an aggregation area for each of Chicago's 50 wards, so that's a good approximate
+     * limit for now.
+     */
+    private static final int MAX_FEATURES = 100;
+    
+    private final FileStorage fileStorage;
+    private final UserPermissions userPermissions;
+    private final String dataSourceId;
+    private final String nameProperty;
+    private final int zoom;
+    private final SpatialDataSource spatialDataSource;
+    private final List<SimpleFeature> finalFeatures;
+
+    // TODO derivations could return their model objects and DataGroups so they don't need direct database and fileStorage access.
+    // A DerivationProduct could be a collection of File, a Collection<M extends BaseModel> and a DataGroup.
+    private final AnalysisCollection<AggregationArea> aggregationAreaCollection;
+    private final AnalysisCollection<DataGroup> dataGroupCollection;
+
+    /**
+     * Extraction, validation and range checking of parameters.
+     * It's kind of a red flag that we're passing Components in here. The products should probably be returned by the
+     * Derivation and stored by some more general purpose wrapper so we can avoid direct file and database access here.
+     * It's also not great to pass in the full request - we only need to extract and validate query parameters.
+     */
+    private AggregationAreaDerivation (FileStorage fileStorage, AnalysisDB database, Request req) {
+
+        // Before kicking off asynchronous processing, range check inputs to fail fast on obvious problems.
+        userPermissions = UserPermissions.from(req);
+        dataSourceId = req.queryParams("dataSourceId");
+        nameProperty = req.queryParams("nameProperty"); //"dist_name"; //
+        zoom = parseZoom(req.queryParams("zoom")); 
+        checkNotNull(dataSourceId);
+        checkNotNull(nameProperty);
```

Comment:
Should nulls be allowed, to indicate the area should be the union of the input features?

Reply:
I have changed this to allow nulls when merging will not happen, but have also added an explicit mergePolygons parameter. It seemed strange to infer something as specific as merging all polygons into one from the simple fact that a particular parameter was not supplied.